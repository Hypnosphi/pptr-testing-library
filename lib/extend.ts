import {extendObjectWithTestingUtils, getDocument} from '.'

const Page = require('puppeteer/lib/Page.js') // tslint:disable-line
const ElementHandle = require('puppeteer/lib/ElementHandle.js') // tslint:disable-line

Page.prototype.document = getDocument

extendObjectWithTestingUtils(ElementHandle.prototype)
