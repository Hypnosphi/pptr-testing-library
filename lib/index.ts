import * as path from 'path'
import {readFileSync} from 'fs'
import {MatcherOptions, Matcher, SelectorMatcherOptions} from 'dom-testing-library/typings/'
import {ElementHandle, EvaluateFn} from 'puppeteer'

const Page = require('puppeteer/lib/Page.js')
const ElementHandle = require('puppeteer/lib/ElementHandle.js')

const domLibraryAsString = readFileSync(
  path.join(__dirname, '../dom-testing-library.js'),
  'utf8',
).replace(/process.env/g, '{}')

const mockFnToExecuteInPage = `
function evaluateInPage(container, fnName, ...args) {
  ${domLibraryAsString}

  const mappedArgs = args.map(item => item.regex ? new RegExp(item.regex) : item)
  return __dom_testing_library__[fnName](container, ...mappedArgs)
}
`

type DOMReturnType = ElementHandle | ElementHandle[] | null

function createDelegateFor(fnName: string): (...args: any[]) => Promise<DOMReturnType> {
  return async function(...args: any[]): Promise<DOMReturnType> {
    if (fnName.includes('All')) throw new Error('*All methods not yet supported')

    // @ts-ignore
    const containerHandle: ElementHandle = this
    // @ts-ignore
    const evaluateFn: EvaluateFn = {toString: () => mockFnToExecuteInPage}

    const mappedArgs = args.map(arg => (arg instanceof RegExp ? {regex: arg.source} : arg))
    const handle = await containerHandle
      .executionContext()
      .evaluateHandle(evaluateFn, containerHandle, fnName, ...mappedArgs)
    const element = handle.asElement()
    if (element) return element
    await handle.dispose()
    return null
  }
}

Page.prototype.getTestingUtilsForDocument = async function(): Promise<ElementHandle> {
  const documentHandle = await this.mainFrame().evaluateHandle('document')
  return await documentHandle.asElement()
}

ElementHandle.prototype.queryByPlaceholderText = createDelegateFor('queryByPlaceholderText')
ElementHandle.prototype.queryAllByPlaceholderText = createDelegateFor('queryAllByPlaceholderText')
ElementHandle.prototype.getByPlaceholderText = createDelegateFor('getByPlaceholderText')
ElementHandle.prototype.getAllByPlaceholderText = createDelegateFor('getAllByPlaceholderText')
ElementHandle.prototype.queryByText = createDelegateFor('queryByText')
ElementHandle.prototype.queryAllByText = createDelegateFor('queryAllByText')
ElementHandle.prototype.getByText = createDelegateFor('getByText')
ElementHandle.prototype.getAllByText = createDelegateFor('getAllByText')
ElementHandle.prototype.queryByLabelText = createDelegateFor('queryByLabelText')
ElementHandle.prototype.queryAllByLabelText = createDelegateFor('queryAllByLabelText')
ElementHandle.prototype.getByLabelText = createDelegateFor('getByLabelText')
ElementHandle.prototype.getAllByLabelText = createDelegateFor('getAllByLabelText')
ElementHandle.prototype.queryByAltText = createDelegateFor('queryByAltText')
ElementHandle.prototype.queryAllByAltText = createDelegateFor('queryAllByAltText')
ElementHandle.prototype.getByAltText = createDelegateFor('getByAltText')
ElementHandle.prototype.getAllByAltText = createDelegateFor('getAllByAltText')
ElementHandle.prototype.queryByTestId = createDelegateFor('queryByTestId')
ElementHandle.prototype.queryAllByTestId = createDelegateFor('queryAllByTestId')
ElementHandle.prototype.getByTestId = createDelegateFor('getByTestId')
ElementHandle.prototype.getAllByTestId = createDelegateFor('getAllByTestId')
ElementHandle.prototype.queryByTitle = createDelegateFor('queryByTitle')
ElementHandle.prototype.queryAllByTitle = createDelegateFor('queryAllByTitle')
ElementHandle.prototype.getByTitle = createDelegateFor('getByTitle')
ElementHandle.prototype.getAllByTitle = createDelegateFor('getAllByTitle')

declare module 'puppeteer' {
  interface Page {
    getTestingUtilsForDocument(): ElementHandle
  }

  interface ElementHandle {
    queryByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle | null>
    queryAllByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    getByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle>
    getAllByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    queryByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle | null>
    queryAllByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    getByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle>
    getAllByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    queryByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
  }
}
