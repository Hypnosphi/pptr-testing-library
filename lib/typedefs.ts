import {ElementHandle} from 'puppeteer'
import {MatcherOptions, Matcher, SelectorMatcherOptions} from 'dom-testing-library/typings/'

type Element = ElementHandle

export interface ITestUtils {
  queryByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByPlaceholderText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  queryByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element | null>
  queryAllByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  getByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element>
  getAllByText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  queryByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element | null>
  queryAllByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  getByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element>
  getAllByLabelText(el: Element, m: Matcher, opts?: SelectorMatcherOptions): Promise<Element[]>
  queryByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByAltText(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  queryByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByTestId(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  queryByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element | null>
  queryAllByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
  getByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element>
  getAllByTitle(el: Element, m: Matcher, opts?: MatcherOptions): Promise<Element[]>
}

declare module 'puppeteer' {
  interface Page {
    getTestingUtilsForDocument(): ElementHandle
  }

  interface ElementHandle {
    queryByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByPlaceholderText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle | null>
    queryAllByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    getByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle>
    getAllByText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    queryByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle | null>
    queryAllByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    getByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle>
    getAllByLabelText(m: Matcher, opts?: SelectorMatcherOptions): Promise<ElementHandle[]>
    queryByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByAltText(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByTestId(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    queryByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle | null>
    queryAllByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
    getByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle>
    getAllByTitle(m: Matcher, opts?: MatcherOptions): Promise<ElementHandle[]>
  }
}
