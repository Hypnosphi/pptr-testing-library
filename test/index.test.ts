import * as path from 'path'
import * as puppeteer from 'puppeteer'
import {getDocument, utils} from '../lib'

describe('lib/index.ts', () => {
  let browser: puppeteer.Browser
  let page: puppeteer.Page

  it('should launch puppeteer', async () => {
    browser = await puppeteer.launch()
    page = await browser.newPage()
    await page.goto(`file://${path.join(__dirname, 'fixtures/page.html')}`)
  })

  it('should export the utilities', async () => {
    const document = await getDocument(page)
    const element = await utils.getByText(document, 'Hello h1')
    expect(await page.evaluate(el => el.textContent, element)).toEqual('Hello h1')
  })

  afterAll(async () => {
    await browser.close()
  })
})
