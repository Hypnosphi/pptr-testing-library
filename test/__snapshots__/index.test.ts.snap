// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/index.ts should handle the LabelText methods 1`] = `"<input id=\\"label-text-input\\" type=\\"text\\">"`;

exports[`lib/index.ts should handle the get* methods 1`] = `"<input type=\\"text\\" data-testid=\\"testid-text-input\\">"`;
